tests/test_directives.py

```diff
@@ -0,0 +1,180 @@
+                        finalize=finalize)
+        asset_dict = {'free': False, 'discount': 5}
+        assert PRODUCT.default == PRODUCT.validation(PRODUCT.default)
+        assert PRODUCT.default == PRODUCT.validation(20)
+        assert PRODUCT.default == 10
+        assert PRODUCT.serial(10, 20) == 10
+        assert PRODUCT.parallel(10, 20) == 10
+        assert PRODUCT.finalize(50, asset_dict) == 45
+        asset_dict['free'] = True
+        assert PRODUCT.finalize(50, asset_dict) == 0
+
+        with pytest.raises(ValueError):
+            PRODUCT.validation(0)
+
+
+class TestDirectives:
+    """
+    Tests for Directives Class
+    """
+    def test_get_directives(self, setUp, available_directives_list):
+        directives = setUp(available_directives_list=available_directives_list)
+        for item in available_directives_list:
+            assert directives[item.name] == item.default
+
+    def test_add_directives(self, setUp, available_directives_list):
+        directives = setUp(available_directives_list=available_directives_list[:-1])
+        directives._add_directive(OMP_NUM_THREADS)
+        assert directives[OMP_NUM_THREADS.name] == OMP_NUM_THREADS.default
+        with pytest.raises(TypeError):
+            directives._add_directive('Test')
+        with pytest.raises(ValueError):
+            directives._add_directive(NP)
+
+    def test_set_directives(self, setUp, available_directives_list):
+        directives = setUp(available_directives_list=available_directives_list[:-1])
+        directives[OMP_NUM_THREADS.name] = OMP_NUM_THREADS.default
+        assert directives[OMP_NUM_THREADS.name] == OMP_NUM_THREADS.default
+
+    def test_define_directive(self, setUp, available_directives_list):
+        directives = setUp(available_directives_list=available_directives_list[:-1])
+        # The values below corresponds to NP, NRANKS, NGPU, EXECUTABLE.
+        valid_values = [100, 100, 100, 'My String']
+        invalid_values = [0, -1, -1, [0]]
+        for idx, dir in enumerate(available_directives_list[:-1]):
+            directives._set_defined_directive(dir.name, valid_values[idx])
+            assert directives[dir.name] == valid_values[idx]
+        for idx, dir in enumerate(available_directives_list[:-1]):
+            with pytest.raises(ValueError):
+                directives._set_defined_directive(dir.name, invalid_values[idx])
+
+    def test_del_directive(self, setUp, available_directives_list):
+        directives = setUp(available_directives_list=available_directives_list[:-1])
+        directives[OMP_NUM_THREADS.name] = OMP_NUM_THREADS.default
+        directives._set_defined_directive(NP.name, 100)
+        assert directives[NP.name] == 100
+        assert directives[OMP_NUM_THREADS.name] == OMP_NUM_THREADS.default
+        del directives[NP.name]
+        assert directives[NP.name] == NP.default
+        del directives[OMP_NUM_THREADS.name]
+        with pytest.raises(KeyError):
```

Comment:
This shouldn't work any more since `PROCESS_FRACTION` is last in the list. This should be tested using a random key not a `_DirectivesItem` used as a user directive.